src/System.Private.CoreLib/shared/System/Text/ASCIIEncoding.cs

```diff
@@ -248,673 +410,338 @@ public override unsafe int GetCharCount(byte[] bytes, int index, int count)
+            if (bytesConsumed != byteCount)
+            {
+                // If there's still data remaining in the source buffer, go down the fallback path.
+                // We need to check for integer overflow since the fallback could change the required
+                // output count in unexpected ways.
 
-            // Just call pointer version
-            int charCount = chars.Length - charIndex;
+                charCount += GetCharCountWithFallback(pBytes, byteCount, bytesConsumed);
+                if (charCount < 0)
+                {
+                    ThrowConversionOverflow();
+                }
+            }
 
-            fixed (byte* pBytes = bytes) fixed (char* pChars = &MemoryMarshal.GetReference((Span<char>)chars))
-                // Remember that charCount is # to decode, not size of array
-                return GetChars(pBytes + byteIndex, byteCount, pChars + charIndex, charCount, null);
+            return charCount;
         }
 
-        // All of our public Encodings that don't use EncodingNLS must have this (including EncodingNLS)
-        // So if you fix this, fix the others.  Currently those include:
-        // EncodingNLS, UTF7Encoding, UTF8Encoding, UTF32Encoding, ASCIIEncoding, UnicodeEncoding
-
-        [CLSCompliant(false)]
-        public unsafe override int GetChars(byte* bytes, int byteCount, char* chars, int charCount)
+        [MethodImpl(MethodImplOptions.AggressiveInlining)] // invoked directly by our common entry method
+        private protected sealed override unsafe int GetCharCountNoFallbackBuffer(byte* pBytes, int byteCount, DecoderFallback fallback, out int bytesConsumed)
         {
-            // Validate Parameters
-            if (bytes == null || chars == null)
-                throw new ArgumentNullException(bytes == null ? nameof(bytes) : nameof(chars), SR.ArgumentNull_Array);
+            Debug.Assert(byteCount >= 0, "Should've been checked by caller.");
 
-            if (charCount < 0 || byteCount < 0)
-                throw new ArgumentOutOfRangeException((charCount < 0 ? nameof(charCount) : nameof(byteCount)), SR.ArgumentOutOfRange_NeedNonNegNum);
+            // See comment in ASCIIEncodingSealed.GetCharCount(char*, int) for details on why we can
+            // perform a short-circuiting computation if we know that we have an DecoderReplacementFallback
+            // whose substitution character is a single BMP char.
 
-            return GetChars(bytes, byteCount, chars, charCount, null);
-        }
+            int numElementsConverted = byteCount;
 
-        public override unsafe int GetChars(ReadOnlySpan<byte> bytes, Span<char> chars)
-        {
-            fixed (byte* bytesPtr = &MemoryMarshal.GetNonNullPinnableReference(bytes))
-            fixed (char* charsPtr = &MemoryMarshal.GetNonNullPinnableReference(chars))
+            if (byteCount != 0 && !(fallback is DecoderReplacementFallback replacementFallback && replacementFallback.MaxCharCount == 1))
```

Comment:
Do you think is it worth it to handle the case of using DecoderExceptionFallback here too? we'll just throw at that time when we hit non ASCII char.

Reply:
I don't think it's a good candidate for optimization. This particular optimization saves us from having to inspect the source input at all, which reduces the runtime of this normally O(n) operation to O(1) instead. For the `DecoderExceptionFallback` case, we'd still need to run through the source input to see if there's any non-ASCII char, which the existing code path does in O(n) time anyway.

Disclaimer: I didn't write the original `EncoderReplacementFallback` / `DecoderReplacementFallback` optimization here. It was in the code before I started mucking about. I left it in because I assumed it was important.